python_modules/dagster/dagster/_core/definitions/declarative_automation/operators/run_operators.py

```diff
@@ -0,0 +1,201 @@
+            return context.root_context.key in (asset_selection | check_selection)
+
+        return await context.asset_graph_view.compute_latest_run_matches_subset(
+            from_subset=context.candidate_subset, filter_fn=_filter_fn
+        )
+
+
+def _run_tag_filter_fn(
+    run_record: "RunRecord",
+    tag_keys: Optional[Set[str]],
+    tag_values: Optional[Mapping[str, str]],
+) -> bool:
+    if tag_keys and not all(key in run_record.dagster_run.tags for key in tag_keys):
+        return False
+    if tag_values and not all(
+        run_record.dagster_run.tags.get(key) == value for key, value in tag_values.items()
+    ):
+        return False
+    return True
+
+
+def _get_run_tag_filter_name(
+    base_name: str,
+    tag_keys: Optional[Set[str]],
+    tag_values: Optional[Mapping[str, str]],
+) -> str:
+    props = []
+    name = base_name
+    if tag_keys is not None:
+        tag_key_str = ",".join(sorted(tag_keys))
+        props.append(f"tag_keys={{{tag_key_str}}}")
+    if tag_values is not None:
+        tag_value_str = ",".join([f"{key}:{value}" for key, value in sorted(tag_values.items())])
+        props.append(f"tag_values={{{tag_value_str}}}")
+
+    if props:
+        name += f"({', '.join(props)})"
+    return name
+
+
+@whitelist_for_serdes
+@record
+class LatestRunExecutedWithTagsCondition(SubsetAutomationCondition):
+    tag_keys: Optional[Set[str]] = None
+    tag_values: Optional[Mapping[str, str]] = None
+
+    @property
+    def name(self) -> str:
+        return _get_run_tag_filter_name("executed_with_tags", self.tag_keys, self.tag_values)
+
+    async def compute_subset(self, context: AutomationContext) -> EntitySubset:  # pyright: ignore[reportIncompatibleMethodOverride]
+        return await context.asset_graph_view.compute_latest_run_matches_subset(
+            from_subset=context.candidate_subset,
+            filter_fn=lambda run_record: _run_tag_filter_fn(
+                run_record, self.tag_keys, self.tag_values
+            ),
+        )
+
+
+@record
+class NewUpdatesWithRunTagsCondition(SubsetAutomationCondition[AssetKey]):
+    tag_keys: Optional[Set[str]] = None
+    tag_values: Optional[Mapping[str, str]] = None
+
+    @property
+    @abstractmethod
+    def base_name(self) -> str: ...
+
+    @property
+    def name(self) -> str:
+        return _get_run_tag_filter_name(self.base_name, self.tag_keys, self.tag_values)
+
+    @abstractmethod
+    def match_candidate_runs(
+        self, candidate_run_ids: Set[str], matching_run_ids: Set[str]
+    ) -> bool: ...
+
+    async def _get_run_records_from_materializations(
+        self,
+        materializations: Sequence["EventLogRecord"],
+        context: LoadingContext,
+    ) -> Sequence["RunRecord"]:
+        from dagster._core.storage.dagster_run import RunRecord
+
+        run_ids = list({record.run_id for record in materializations if record.run_id})
+        if not run_ids:
+            return []
```

Comment:
extremely nit-picky but technically we don't need this right? particularly with the updated implementation of `_blocking_batch_load` on RunRecord, if there are no keys passed in, then no query will be executed.

also we guard against no new materializations in the `compute_subset` method, so it should actually be possible to get here without at least one run_id

if you have even the slightest desire to keep this guard, im totally fine with that though